main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala

```diff
@@ -116,4 +116,96 @@ object SafetyError {
     })
   }
 
+  /**
+    * An error raised to indicate an illegal object derivation. Objects can only be derived from Java interfaces.
+    *
+    * @param clazz The (illegal) superclass.
+    * @param loc   The source location of the object derivation.
+    */
+  case class IllegalObjectDerivation(clazz: java.lang.Class[_], loc: SourceLocation) extends SafetyError {
+    def summary: String = s"Illegal object derivation from '${clazz}'."
+
+    def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+        |>> '${red(clazz.toString)}' is not a Java interface.
+        |
+        |${code(loc, "the illegal derivation occurs here.")}
+        |""".stripMargin
+    }
+
+    def explain(formatter: Formatter): Option[String] = None
+  }
+
+  /**
+    * An error raised to indicate an invalid `this` parameter for a method.
+    *
+    * @param clazz The Java interface being implemented.
+    * @param name  The name of the method with the invalid `this` parameter.
+    * @param loc   The source location of the method.
+    */
+  case class InvalidThis(clazz: java.lang.Class[_], name: String, loc: SourceLocation) extends SafetyError {
```

Comment:
I assume `clazz` is the actual class, thus we should also add the declared type. Then we give a hint as shown below.